zoo of create functions (#2895)
vyperlang/vyper: docs/built-in-functions.rst

.. note::

  There is no runtime check that there is code already deployed at ``target`` (since a proxy may be deployed counterfactually). Most applications may want to insert this check.

.. note::

  In previous versions of vyper, this function was named ``create_forwarder_to``.

.. py:function:: create_copy_of(target: address, value: uint256 = 0[, salt: bytes32]) -> address

    Create a physical copy of the runtime code at ``target``. The code at ``target`` is byte-for-byte copied into a newly deployed contract.

    * ``target``: Address of the contract to copy
    * ``value``: The wei value to send to the new contract address (Optional, default 0)
    * ``salt``: A ``bytes32`` value utilized by the deterministic ``CREATE2`` opcode (Optional, if not supplied, ``CREATE`` is used)

    Returns the address of the created contract. If the create operation fails (for instance, in the case of a ``CREATE2`` collision), execution will revert. If there is no code at ``target``, execution will revert.

    .. code-block:: python

        @external
        def foo(_target: address) -> address:
            return create_copy_of(_target)

> .. py:function:: create_with_code_of(target: address, *args, value: uint256 = 0[, salt: bytes32]) -> address

Reviewer: hmm, maybe `create_from_factory` might make more sense?

Author: i was thinking about this, it makes sense, especially if we can standardize on the "factory" terminology

Author: `create_with_factory_at()`?

Reviewer: Totally better yeah

Author: renamed to `create_from_factory()`

>     Copy the code of ``target`` into memory and execute it as initcode. In other words, this operation interprets the code at ``target`` not as regular runtime code, but as a factory contract. The *args are interpreted as constructor arguments, and are ABI-encoded and included when executing the initcode.

Reviewer: ```suggestion
    Copy the code of ``target`` into memory and execute it as initcode. In other words, this operation interprets the code at ``target`` not as regular runtime code, but as a factory contract. The ``*args`` are interpreted as constructor arguments, and are ABI-encoded and included when executing the initcode.
```

    * ``target``: Address of the contract to proxy to
    * ``value``: The wei value to send to the new contract address (Optional, default 0)
    * ``salt``: A ``bytes32`` value utilized by the deterministic ``CREATE2`` opcode (Optional, if not supplied, ``CREATE`` is used)

    Returns the address of the newly created proxy contract. If the create operation fails (for instance, in the case of a ``CREATE2`` collision), execution will revert.

    .. code-block:: python

        @external
        def foo(_target: address) -> address:
            return create_minimal_proxy_to(_target)

.. note::

  It is very important that the deployed contract at ``target`` is code you know and trust, and does not implement the ``selfdestruct`` opcode or have upgradeable code as this will affect the operation of the proxy contract.

.. note::

  There is no runtime check that there is code already deployed at ``target`` (since a proxy may be deployed counterfactually). Most applications may want to insert this check.

.. note::

>   In previous versions of vyper, this function was named ``create_forwarder_to``.

Reviewer 2: ```suggestion
  Before version 0.3.4, this function was named ``create_forwarder_to``.
```

.. py:function:: create_copy_of(target: address, value: uint256 = 0[, salt: bytes32]) -> address

    Create a physical copy of the runtime code at ``target``. The code at ``target`` is byte-for-byte copied into a newly deployed contract.

    * ``target``: Address of the contract to copy
    * ``value``: The wei value to send to the new contract address (Optional, default 0)
    * ``salt``: A ``bytes32`` value utilized by the deterministic ``CREATE2`` opcode (Optional, if not supplied, ``CREATE`` is used)

    Returns the address of the created contract. If the create operation fails (for instance, in the case of a ``CREATE2`` collision), execution will revert. If there is no code at ``target``, execution will revert.

    .. code-block:: python

        @external
        def foo(_target: address) -> address:
            return create_copy_of(_target)

.. note::

>     The implementation of ``create_copy_of`` assumes that the code at ``target`` is smaller than 16MB. While this is much larger than the EIP-170 constraint of 24KB, it is a conservative size limit intended to future proof deployer contracts in case the EIP-170 constraint is lifted. If the code at ``target`` is larger than 16MB, the behavior of ``create_copy_of`` is undefined.

Reviewer 2: ```suggestion
    The implementation of ``create_copy_of`` assumes that the code at ``target`` is smaller than 16MB. While this is much larger than the EIP-170 constraint of 24KB, it is a conservative size limit intended to future-proof deployer contracts in case the EIP-170 constraint is lifted. If the code at ``target`` is larger than 16MB, the behavior of ``create_copy_of`` is undefined.
```
